src/blas/impl/KokkosBlas2_gemv_impl.hpp

```diff
@@ -479,6 +479,175 @@ singleLevelGemv (const char trans[],
   }
 }
 
+
+// ---------------------------------------------------------------------------------------------
+// Functor for a two-level parallel_reduce version of (conjugate)
+// transpose GEMV.  The functor uses parallel-for over the columns of the input
+// matrix A and each team uses parallel-reduce over the row of its column.
+// The output vector y is the reduction result.
+template<class AViewType,
+         class XViewType,
+         class YViewType,
+         const bool conj,
+         class IndexType = typename AViewType::size_type>
+struct TwoLevelTransposeGEMV {
+  typedef typename YViewType::non_const_value_type y_value_type;
+  typedef typename AViewType::non_const_value_type AlphaCoeffType;
+  typedef typename YViewType::non_const_value_type BetaCoeffType;
+
+
+  typedef typename AViewType::execution_space execution_space;
+  typedef Kokkos::TeamPolicy<execution_space> policy_type;
+  typedef typename policy_type::member_type member_type;
+
+  TwoLevelTransposeGEMV (const AlphaCoeffType& alpha,
+                         const AViewType& A,
+                         const XViewType& x,
+                         const BetaCoeffType& beta,
+                         const YViewType& y) :
+    alpha_ (alpha), A_ (A), x_ (x), beta_ (beta), y_ (y)
+  {
+    static_assert (Kokkos::Impl::is_view<AViewType>::value,
+                   "AViewType must be a Kokkos::View.");
+    static_assert (Kokkos::Impl::is_view<XViewType>::value,
+                   "XViewType must be a Kokkos::View.");
+    static_assert (Kokkos::Impl::is_view<YViewType>::value,
+                   "YViewType must be a Kokkos::View.");
+    static_assert (static_cast<int> (AViewType::rank) == 2,
+                   "AViewType must have rank 2.");
+    static_assert (static_cast<int> (XViewType::rank) == 1,
+                   "XViewType must have rank 1.");
+    static_assert (static_cast<int> (YViewType::rank) == 1,
+                   "YViewType must have rank 1.");
+    static_assert (std::is_integral<IndexType>::value,
+                   "IndexType must be an integer.");
+  }
+
+public:
+  KOKKOS_INLINE_FUNCTION void
+  operator () (const member_type & team) const
+  {
+    using Kokkos::Details::ArithTraits;
+    typedef ArithTraits<typename AViewType::non_const_value_type> KAT;
+
+    const IndexType M = A_.extent(0);
+    const int j = team.league_rank(); // batch id
+
+    // parallel-reduce to compute val += A(:,j)' * x
+    y_value_type val = KAT:: zero();
+    Kokkos::parallel_reduce( Kokkos::TeamThreadRange( team, M ), [&] ( const int i, y_value_type &update ) {
+      const auto x_i = x_(i);
+      const auto A_ij = conj ? KAT::conj (A_(i,j)) : A_(i,j);
```

Comment:
Will this have the same issue with Sacado types as the recent "unary `operator-` in a trinary operator" issue?

Reply:
@mhoemmen That's a good question... Eric pushed some changes to Sacado to mark the relevant expression template-related unary ctors `explicit` in hopes of avoiding those issues trilinos/Trilinos#6411, should a preemptive change be made here just in case?

Reply:
@ndellingwood I think there are other places in kokkos-kernels that do this, so if it builds and passes tests in Trilinos, perhaps we shouldn't worry so much.  The "trinary operator with conjugate" case is much more common in kokkos-kernels I suspect.